src/middlewared/middlewared/plugins/vm.py

```diff
@@ -1345,6 +1341,126 @@ def decompress_gzip(self, src, dst):
         return vnc_web
 
 
+class VMDeviceService(CRUDService):
+
+    class Config:
+        namespace = 'vm.device'
+        datastore = 'vm.device'
+        datastore_extend = 'vm.device.extend_device'
+
+    @private
+    async def extend_device(self, device):
+        device['vm'] = device['vm']['id']
+        if device['order'] is None:
+            if device['dtype'] == 'CDROM':
+                device['order'] = 1000
+            elif device['dtype'] in ('DISK', 'RAW'):
+                device['order'] = 1001
+            else:
+                device['order'] = 1002
+        return device
+
+    @accepts(
+        Dict(
+            'vmdevice_create',
+            Str('dtype', enum=['NIC', 'DISK', 'CDROM', 'VNC', 'RAW'], required=True),
+            Int('vm'),
+            Dict('attributes', additional_attrs=True),
+            Int('order', default=None),
+            register=True,
+        ),
+    )
+    async def do_create(self, data):
+
+        if not data.get('vm'):
+            raise ValidationError('vmdevice_create.vm', 'This field is required.')
+
+        await self.validate_device(data)
+        id = await self.middleware.call('datastore.insert', self._config.datastore, id, data)
+        await self.__reorder_devices(id, data['vm'], data['order'])
+
+        return await self._get_instance(id)
+
+    @accepts(Int('id'), Patch(
+        'vmdevice_create',
+        'vmdevice_update',
+        ('attr', {'update': True}),
+    ))
+    async def do_update(self, id, data):
+        device = await self._get_instance(id)
+        new = device.copy()
+        new.update(data)
+
+        await self.validate_device(new, device)
+        await self.middleware.call('datastore.update', self._config.datastore, id, new)
+        await self.__reorder_devices(id, device['vm'], new['order'])
+
+        return await self._get_instance(id)
+
+    @accepts(Int('id'))
+    async def do_delete(self, id):
+        await self.middleware.call('datastore.delete', self._config.datastore, id)
+
+    async def __reorder_devices(self, id, vm_id, order):
+        if order is None:
+            return
+        filters = [('vm', '=', vm_id), ('id', '!=', id)]
+        if await self.middleware.call('vm.device.query', filters + [
+            ('order', '=', order)
+        ]):
+            used_order = [order]
+            for device in await self.middleware.call(
+                'vm.device.query', filters, {'order_by': ['order']}
+            ):
+                if device['order'] is None:
+                    continue
+
+                if device['order'] not in used_order:
+                    used_order.append(device['order'])
+                    continue
+
+                device['order'] = min(used_order) + 1
+                while device['order'] in used_order:
```

Comment:
Wouldn't this always be the case? Where does it get removed from the array?

Reply:
I dont think so? Only devices seen are put in used_order, they dont need to be removed, I only need order that was not used by anything else.

We need to change order of everything else, e.g. if I had devices with orders [0, 1, 2, 3] and I update/add a new device of order 0, all devices will be updated: [0, 1, 2, 3, 4]